splice_machine/src/main/java/com/splicemachine/derby/stream/function/merge/AbstractMergeJoinFlatMapFunction.java

```diff
@@ -232,10 +416,137 @@ else if (!leftKeyIsNull) {
             if (scanInfo != null && scanInfo.getSameStartStopPosition())
                 ((BaseActivation)joinOperation.getActivation()).setScanStopOverride(startPosition);
         }
+        return true;
+    }
+
+    private void startNewRightSideScan(PeekingIterator<ExecRow> leftRows) throws StandardException {
+        Iterator<ExecRow> rightIterator;
+
+        ArrayList<Pair<ExecRow, ExecRow>> keyRows = null;
+
+        boolean skipRightSideRead = false;
+
+        // The mem platform doesn't support the HBase MultiRangeRowFilter.
+        if (!IS_MEM_PLATFORM && leftRows instanceof BufferedMergeJoinIterator) {
+            BufferedMergeJoinIterator mjIter = (BufferedMergeJoinIterator)leftRows;
+            keyRows = mjIter.getKeyRows();
+            ((BaseActivation) joinOperation.getActivation()).setKeyRows(keyRows);
+            skipRightSideRead = (keyRows == null);
+        }
+        else
+            skipRightSideRead = !initRightScanWithStartStopKeys(leftRows);
+
+        // If there are no join keys to look up in the right table,
+        // don't even read the right table.
+        if (skipRightSideRead)
+            rightIterator = Collections.emptyIterator();
+        else {
+            rightSide = joinOperation.getRightOperation();
+            rightSide.reset();
```

Comment:
You're right.  The rightSide.close() calls don't actually get applied until the stream of left rows is complete, so the call to reset can be removed.  I will remove it.

Reply:
It turns out when the merge join is the source of the right side of a nested loop join it may close the operation and then restart it.  I added in a new method called reOpen, which does not effect the other fields like badRecords.